### PR TITLE
Let TreeEditor.ModelService implementation be asynchronous #21

### PR DESCRIPTION
**Pull request by vhemery**
getDataForNode, getSchemaForNode and getUiSchemaForNode
can now have asynchronous implementation.
getChildrenMapping and getNameForType
can not be made asynchronous because of the theia core interfaces.

Signed-off-by: vhemery <vincent.hemery@csgroup.eu>

### CODE REVIEW

File: theia-tree-editor/src/browser/tree-editor-widget.tsx
Comment (lucas-koehler) [inline theia-tree-editor/src/browser/tree-editor-widget.tsx:124]: I'd just make this a normal `Promise`. It makes sense that this is handled asynchronously anyway.

File: theia-tree-editor/src/browser/tree-editor-widget.tsx
Comment (lucas-koehler) [inline theia-tree-editor/src/browser/tree-editor-widget.tsx:136]: I'd just make this a normal `Promise`. It makes sense that this is handled asynchronously anyway.

File: theia-tree-editor/src/browser/tree-editor-widget.tsx
Comment (lucas-koehler) [inline theia-tree-editor/src/browser/tree-editor-widget.tsx:164]: I'd just make this a normal `Promise`. It makes sense that this is handled asynchronously anyway.

File: theia-tree-editor/src/browser/master-tree-widget.tsx
Comment (lucas-koehler) [inline theia-tree-editor/src/browser/master-tree-widget.tsx:317]: Please re-write this to be more readable and avoid shadowing variable names: `data.map((d, i)` shadows `const { d, i }`.
E.g. something like this:
```
// d = data; i = index
const indexedData = data.map((d, i) => ({ d, i }));
for (const { d, i } of indexedData) {
  await this.nodeFactory.mapData(d, node, property, index + i);
}
```

File: theia-tree-editor/src/browser/interfaces.ts
Comment (lucas-koehler) [inline theia-tree-editor/src/browser/interfaces.ts:86]: I'd change the return type to `MaybePromise<JsonSchema | undefined>` because the Promise can also evaluate to `undefined`.

File: theia-tree-editor/src/browser/interfaces.ts
Comment (lucas-koehler) [inline theia-tree-editor/src/browser/interfaces.ts:95]: I'd change the return type to `MaybePromise<UISchemaElement | undefined>` because the Promise can also evaluate to `undefined`.

**Review by lucas-koehler — CHANGES_REQUESTED** [review]
Hello @vhemery. First of all, thank you very much for your contribution :)

In general, this looks good to me but I just have some minor comments.
Before you fix these, you might want to wait for https://github.com/eclipse-emfcloud/theia-tree-editor/pull/24 to be merged because this adapts `theia-tree-editor/src/browser/detail-form-widget.tsx` quite heavily and will cause conflicts. As this is a more complicated rework than the async methods, I'd like to merge that first.

So I suggest you either wait until it's merged or just rebase these changes on the PR's branch (it's part of this repository) :)

File: theia-tree-editor/src/browser/tree-editor-widget.tsx
Comment (vhemery) [reply theia-tree-editor/src/browser/tree-editor-widget.tsx:124]: The big advantage of MaybePromise is to support legacy implementations without having to change them.
The coffee editor is a good well-known example, and I don't think it would be impacted with MaybePromise.
https://github.com/eclipse-emfcloud/coffee-editor/blob/master/web/coffee-editor-extension/src/browser/coffee-tree/coffee-tree-editor-widget.tsx
On the other hand, Promise would force it and all users who used it as a reference implementation to update their code.

It doesn't bother me to change ResourceTreeEditorWidget which provides a default implementation probably not overridden, but forcing all other implementers to change their implementation from sync to async looks like a big deal to me.
(this same conversation is valid for all similar MaybePromise => Promise suggestions)

Comment (vhemery) [reply theia-tree-editor/src/browser/tree-editor-widget.tsx:124]: OK for changes.

**Comment by vhemery** [comment]
Continued in #29

**Pull request closed without merging**
